doc/kopano-admin.8
.TH "KOPANO\-ADMIN" "8" "November 2016" "Kopano 8" "Kopano Core user reference"
.\" http://bugs.debian.org/507673
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
kopano-admin \- Manages Kopano users and stores.
.SH "SYNOPSIS"
.HP \w'\fBkopano\-admin\fR\ 'u
\fBkopano\-admin\fR \fIACTION\fR [\fIOPTION\fR...]
.SH "DESCRIPTION"
.PP
This tool can be used to create the public store and to add, update and remove users from Kopano. The storage server must be running for
\fBkopano\-admin\fR
to work.
.PP
If no action is given, a listing of the possible parameters is printed. When invalid actions or not enough options for an action are given, an error message is printed.
.PP
When using LDAP as the users source, create, modify and delete actions are done in the LDAP tree and not using the kopano\-admin tool. Please see the
\fBEXTERNAL USERS\fR()
section for more information.
.SH "ACTIONS"
.PP
kopano\-admin needs an action command with the appropriate options. Valid actions are:
.PP
\fB\-s\fR
.RS 4
Create a public store. No other options are needed. Only one public store can be created. Successive calls will fail.
.RE
.PP
\fB\-c\fR \fIusername\fR
.RS 4
Create a new user,
\fB\-p\fP,
\fB\-f\fP
and
\fB\-e\fP
options are required,
\fB\-a\fP
and
\fB\-n\fP
parameters are optional. To set a password using a password prompt, use the
\fB\-P\fP
option instead of
\fB\-p\fP.
.RE
.PP
\fB\-d\fR \fIusername\fR
.RS 4
Delete a user. No other options are needed. The deleted store of the user will be marked as orphan store and can be restored with
\fB\-\-hook\-store\fP.
.RE
.PP
\fB\-u\fR \fIusername\fR
.RS 4
Update user information. Valid parameters are:
[\fB\-p\fP|\fB\-P\fP],
\fB\-f\fP,
\fB\-e\fP,
\fB\-a\fP,
\fB\-n\fP
and
\fB\-U\fP
to update user information. Use:
\fB\-\-qo\fP,
\fB\-\-qw\fP,
\fB\-\-qs\fP
or
\fB\-\-qh\fP
to set quota levels. Use 0 with quota options to set as unlimited.
.sp
\fB\-\-enable\-feature\fP
and
\fB\-\-disable\-feature\fP
to enable or disable specific features for users.
.sp
This action is also required for all options to control the autoaccept of meeting requests. See the
\fB\-\-mr\-accept\fP,
\fB\-\-mr\-decline\-conflict\fP,
\fB\-\-mr\-decline\-recurring\fP and \fB\-\-mr\-process\fP
options.
.RE
.PP
\fB\-g\fR \fIgroupname\fR
.RS 4
Create a new group. Valid parameters are:
\fB\-e\fP
.RE
.PP
\fB\-\-update\-group\fR \fIgroupname\fR
.RS 4
Update group information. Valid parameters are:
\fB\-e\fP
.RE
.PP
\fB\-G\fR \fIgroupname\fR
.RS 4
Delete a group. No other options are needed.
.RE
.PP
\fB\-b\fR \fIusername\fR
.RS 4
Add a user to a group. Use the
\fB\-i\fP
to set the groupname.
.RE
.PP
\fB\-B\fR \fIusername\fR
.RS 4
Remove a user from a group. Use the
\fB\-i\fP
to set the groupname.
.RE
.PP
\fB\-l\fR, \fB\-\-list\-users\fP
.RS 4
List all users available in Kopano. When using an external user source, this action will implicitly synchronize all users in the external source, creating, updating and/or removing users and stores.
.RE
.PP
\fB\-L\fR, \fB\-\-list\-groups\fP
.RS 4
List all groups available in Kopano. When using an external user source, this action will implicitly synchronize all groups in the external source, creating updating and/or removing groups and memberships.
.RE
.PP
\fB\-\-list\-companies\fR
.RS 4
List all tenans available in Kopano. When using an external user source, this action will implicitly synchronize all tenants in the external source, creating updating and/or removing companies.
.sp
This option is only available in multi-tenancy Kopano
.RE
.PP
\fB\-\-details\fR \fIname\fR
.RS 4
Show all the details of a user, showing the fullname, e-mailaddress, active state, administator state, group memberships and quota settings. Optionally use
\fB\-\-type\fP
to indicate for what kind of object the details are being requested.
.sp
Note: This function does not synchronize with the external user plugin. Thus changes from e.g. LDAP will not be set during this function.
.RE
.PP
\fB\-\-type\fR \fItype\fR
.RS 4
Additional argument for
\fB\-\-details\fP. The argument with this option indicates for what type of object the details are being requested. Allowed values are
\fBuser\fP,
\fBgroup\fP
or
\fBcompany\fP
When this option is not used, it defaults to
\fBuser\fP.
.RE
.PP
\fB\-\-create\-company\fR \fIcompanyname\fR
.RS 4
Create a new tenant space.
.sp
Use:
\fB\-\-qo\fP,
\fB\-\-qw\fP,
\fB\-\-qs\fP,
\fB\-\-qh\fP
to set quota levels for the tenant.
.sp
Use:
\fB\-\-udqo\fP,
\fB\-\-udqw\fP,
\fB\-\-udqs\fP,
\fB\-\-udqh\fP
to set the default quota levels for the users inside the tenant.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-update\-company\fR \fIcompanyname\fR
.RS 4
Update an existing tenant space.
.sp
Use:
\fB\-\-qo\fP,
\fB\-\-qw\fP,
\fB\-\-qs\fP,
\fB\-\-qh\fP
to set quota levels for the tenant.
.sp
Use:
\fB\-\-udqo\fP,
\fB\-\-udqw\fP,
\fB\-\-udqs\fP,
\fB\-\-udqh\fP
to set the default quota levels for the users inside the .
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-delete\-company\fR \fIcompanyname\fR
.RS 4
Delete company space.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-set\-system\-admin\fR \fIusername\fP
.RS 4
Set system administrator for the tenant specified by
\fB\-I\fP.
.sp
Please be aware that this option does not provide the user with administrator privileges. The system administrator is considered the main contact person for a company, it will for example be used as default sender for quota warning emails.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-add\-to\-viewlist\fR \fIcompanyname\fR
.RS 4
Add tenant
\fIcompanyname\fP
to remote view list of company specified by
\fB\-I\fP. After this command the
\fIcompanyname\fP
is capable of viewing all members of the company specified by
\fB\-I\fP.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-del\-from\-viewlist\fR \fIcompanyname\fR
.RS 4
Delete company
\fIcompanyname\fP
from remote view list of tenant specified by
\fB\-I\fP. After this command the
\fIcompanyname\fP
is no longer capable of viewing all members of the tenant specified by
\fB\-I\fP.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-list\-view\fR
.RS 4
List all tenants in the remote view list of the tenant specified by
\fB\-I\fP. The tenants in this list are able to view all members of the specified tenant in their Address Book.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-add\-to\-adminlist\fR \fIusername\fR
.RS 4
Add user
\fIusername\fP
to remote\-admin list of tenant specified by
\fB\-I\fP. This is the administrator list for remote administrators, as such it only manages administrators from a different tenant. Users who should be administrator over their own tenant are managed by updating (\fB\-u\fR) the user and specifying the
\fB\-a\fR
argument.
.sp
Users can only be administrator over a different company when they have also been granted view privileges, can be granted by using the
\fB\-\-add\-to\-viewlist\fR.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-del\-from\-adminlist\fR \fIusername\fR
.RS 4
Delete user
\fIusername\fP
from remote\-admin list of company specified by
\fB\-I\fP. This is the administrator list for remote administrators, as such it only manages administrators from a different tenant. Users who should be administrator over their own tenant are managed by updating (\fB\-u\fR) the user and specifying the
\fB\-a\fR
argument.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-list\-admin\fR
.RS 4
List all users in the remote\-admin list of the tenant specified by
\fB\-I\fP. This is the administrator list for remote administrators, as such it only manages administrators from a different tenant. Users who should be administrator over their own tenant are managed by updating (\fB\-u\fR) the user and specifying the
\fB\-a\fR
argument.
.sp
Users can only be administrator over a different tenant when they have also been granted view privileges, can be granted by using the
\fB\-\-add\-to\-viewlist\fR.
.sp
This option is only available in multi-tenancy Kopano.
.RE
.PP
\fB\-\-add\-userquota\-recipient\fR \fIuser\fR
.RS 4
Add
\fIuser\fP
as recipient to userquota warning emails. You must use
\fB\-I\fP
to set the tenant space to apply the recipient action on.
.RE
.PP
\fB\-\-del\-userquota\-recipient\fR \fIuser\fR
.RS 4
Delete
\fIuser\fP
as recipient to userquota warning emails. You must use
\fB\-I\fP
to set the tenant space to apply the recipient action on.
.RE
.PP
\fB\-\-list\-userquota\-recipients\fR
.RS 4
List all additional recipients for a userquota warning email. Use
\fB\-I\fP
to request the recipient list for a particular tenant space.
.RE
.PP
\fB\-\-add\-companyquota\-recipient\fR \fIuser\fR
.RS 4
Add
\fIuser\fP
as recipient to tenant quota warning emails. You must use
\fB\-I\fP
to set the tenant space to apply the recipient action on.
.RE
.PP
\fB\-\-del\-companyquota\-recipient\fR \fIuser\fR
.RS 4
Delete
\fIuser\fP
as recipient to tenant quota warning emails. You must use
\fB\-I\fP
to set the tenant space to apply the recipient action on.
.RE
.PP
\fB\-\-list\-companyquota\-recipients\fR
.RS 4
List all additional recipients for a tenant quota warning email. Use
\fB\-I\fP
to request the recipient list for a particular tenant space
.RE
.PP
\fB\-\-list\-sendas\fR \fIuser\fR
.RS 4
List all users who are able to directly send an email as
\fIuser\fR. This has been set in the LDAP server, or with the
\fB\-\-add\-sendas\fR
command for Unix and DB plugins. Optionally use
\fB\-\-type\fP
to indicate for what kind of object the sendas details are being requested.
.RE
.PP
\fB\-\-clear\-cache\fR
.RS 4
Clears the server's caches. All data cached inside the kopano\-server is cleared. Although this can never cause any data loss, it can affect the performance of your server, since any data requested after the cache is cleared needs to be re\-requested from the database or LDAP server. Normally this option is never needed; it is mostly used as a diagnostics tool.
.sp
Optionally use
\fB\-\-clear\-cache=\fP
to specify a set of purge options. The following options can be used:
.RS 4
0x0000    Release no longer used memory back to the kernel
.RE
.RS 4
0x0001    Purge the quota cache
.RE
.RS 4
0x0002    Purge the quota default cache
.RE
.RS 4
0x0004    Purge the object cache
.RE
.RS 4
0x0008    Purge the store cache
.RE
.RS 4
0x0010    Purge the ACL cache
.RE
.RS 4
0x0020    Purge the cell cache
.RE
.RS 4
0x0040    Purge the index1 cache
.RE
.RS 4
0x0080    Purge the index2 cache
.RE
.RS 4
0x0100    Purge the indexproperty cache
.RE
.RS 4
0x0200    Purge the user object cache
.RE
.RS 4
0x0400    Purge the externid cache
.RE
.RS 4
0x0800    Purge the userdetails cache
.RE
.RS 4
0x1000    Purge the server cache
.RE
.RE
.PP
\fB\-\-purge\-softdelete\fR \fIdays\fR
.RS 4
Starts a softdelete purge on the server, removing all soft\-deleted items which have been deleted
\fIdays\fR
days ago, or earlier
.RE
.PP
\fB\-\-purge\-deferred\fR
.RS 4
The server has an optimization in which changes to the tproperties table are
not writted directly, but delayed for a more efficient write at a later time.
The server auto\-purges these regularly. This command allows you to purge all
changes pending. It may be useful to run this during low I/O load of your
server (e.g. at night).
.RE
.PP
\fB\-\-list\-orphans\fR
.RS 4
When a user is removed, the store becomes orphaned. This option shows a list of stores that are not hooked to a user. You can use the
\fB\-\-remove\-store\fR
and
\fB\-\-hook\-store\fR
from this list.
.RE
.PP
\fB\-\-hook\-store\fR \fIstore\-guid\fR
.RS 4
You can hook an orphaned store to an existing user, so you may access the store again. Use the
\fB\-u\fR
\fIusername\fR
to specify the user to hook the store to.
.sp
You can copy an orphaned store to the public store. Use the
\fB\-\-copyto\-public\fR
to copy the store to the public folder named "Deleted Stores". This folder is then only visible for users with Kopano admin privileges.
.sp
To hook a public store, use
\fB\-\-type\fR
group/company to influence the name type in the \fB\-u\fP switch.
.sp
To hook an archive store, use
\fB\-\-type\fR
\fIarchive\fR.
.RE
.PP
\fB\-\-remove\-store\fR \fIstore\-guid\fR
.RS 4

swig/python/kopano/scripts/kopano-admin.py
#!/usr/bin/env python
import getpass
import locale
import sys
import traceback

from MAPI.Tags import PR_EC_STATSTABLE_SYSTEM, PR_DISPLAY_NAME, PR_EC_STATS_SYSTEM_VALUE, PR_EC_RESYNC_ID
import kopano
from kopano.parser import _true, _int, _name, _guid, _bool, _list_name, _date

def parser_opt_args():
    parser = kopano.parser('skpcuGCf')
    parser.add_option('--debug', help='Debug mode', **_true())
    parser.add_option('--lang', help='Create folders in this language')
    parser.add_option('--sync', help='Synchronize users and groups with external source', **_true())
    parser.add_option('--clear-cache', help='Clear all caches in the server', **_true())
    parser.add_option('--purge-softdelete', help='Purge items in marked as softdeleted that are older than N days', **_int())
    parser.add_option('--purge-deferred', help='Purge all items in the deferred update table', **_true())
    parser.add_option('--list-users', help='List users', **_true())
    parser.add_option('--list-groups', help='List groups', **_true())
    parser.add_option('--list-companies', help='List companies', **_true())
    parser.add_option('--list-orphans', help='List orphan stores', **_true())
    parser.add_option('--user-count', help='Output the system user counts', **_true())
    parser.add_option('--remove-store', help='Remove orphaned store', **_guid())
    parser.add_option('--create', help='Create object', **_true())
    parser.add_option('--delete', help='Delete object', **_true())
    parser.add_option('--name', help='Name', **_name())
    parser.add_option('--fullname', help='Full name', **_name())
    parser.add_option('--email', help='Email address', **_name())
    parser.add_option('--password', help='Password', **_name())
    parser.add_option('--password-prompt', help='Password (prompt)', **_true())
    parser.add_option('--create-store', help='Create store', **_true())
    parser.add_option('--unhook-store', help='Unhook (public) store', **_true())
    parser.add_option('--unhook-archive', help='Unhook archive store', **_true())
    parser.add_option('--hook-store', help='Hook store', **_guid())
    parser.add_option('--hook-archive', help='Hook archive store', **_guid())
    parser.add_option('--reset-folder-count', help='Reset folder counts', **_true())
    parser.add_option('--add-companyquota-recipient', help='User to add to companyquota recipients', **_list_name())
    parser.add_option('--remove-companyquota-recipient', help='User to remove from companyquota recipients', **_list_name())
    parser.add_option('--add-userquota-recipient', help='User to add to userquota recipients', **_list_name())
    parser.add_option('--remove-userquota-recipient', help='User to remove from userquota recipients', **_list_name())
    parser.add_option('--active', help='User is active', **_bool())
    parser.add_option('--mr-accept', help='Auto-accept meeting requests', **_bool())
    parser.add_option('--mr-accept-conflicts', help='Auto-accept conflicting meeting requests', **_bool())
    parser.add_option('--mr-accept-recurring', help='Auto-accept recurring meeting requests', **_bool())
    parser.add_option('--add-feature', help='Feature to add', **_list_name())
    parser.add_option('--remove-feature', help='Feature to remove', **_list_name())
    parser.add_option('--admin', help='Administrator', **_name())
    parser.add_option('--add-admin', help='User to add as remote-admin', **_list_name())
    parser.add_option('--remove-admin', help='User to remove as remote-admin', **_list_name())
    parser.add_option('--add-view', help='Company to add as remote-viewer of address book', **_list_name())
    parser.add_option('--remove-view', help='Company to remove as remote-viewer of address book', **_list_name())
    parser.add_option('--admin-level', help='Admin level', **_int())
    parser.add_option('--add-sendas', help='User to add to send-as', **_list_name())
    parser.add_option('--remove-sendas', help='User to remove from send-as', **_list_name())
    parser.add_option('--add-delegate', help='User to add to delegation', **_list_name())
    parser.add_option('--remove-delegate', help='User to remove from delegation', **_list_name())
    parser.add_option('--add-permission', help='Permission to add (member:right1,right2..)', **_list_name())
    parser.add_option('--remove-permission', help='Permission to remove', **_list_name())
    parser.add_option('--add-user', help='User to add', **_list_name())
    parser.add_option('--remove-user', help='User to remove', **_list_name())
    parser.add_option('--quota-override', help='Override server quota limits', **_bool())
    parser.add_option('--quota-hard', help='Hardquota limit in MB', **_int())
    parser.add_option('--quota-soft', help='Softquota limit in MB', **_int())
    parser.add_option('--quota-warn', help='Warnquota limit in MB', **_int())
    parser.add_option('--ooo-active', help='Out-of-office is active', **_bool())
    parser.add_option('--ooo-clear', help='Clear Out-of-office settings', **_true())
    parser.add_option('--ooo-subject', help='Out-of-office subject', **_name())
    parser.add_option('--ooo-message', help='Out-of-office message (file)', **_name())
    parser.add_option('--ooo-from', help='Out-of-office from date', **_date())
    parser.add_option('--ooo-until', help='Out-of-office until date', **_date())

    return (parser,) + parser.parse_args()

ACTION_MATRIX = {
    ('list_users',): ('global', 'companies', 'groups'),
    ('list_groups',): ('global', 'companies',),
    ('create', 'delete'): ('companies', 'groups', 'users'),
    ('sync', 'clear_cache', 'purge_softdelete', 'purge_deferred', 'remove_store'): ('global',),
    ('list_orphans', 'list_companies', 'user_count'): ('global',),
}

UPDATE_MATRIX = {
    ('name',): ('companies', 'groups', 'users'),
    ('email', 'fullname', 'add_sendas', 'remove_sendas'): ('users', 'groups'),
    ('password', 'password_prompt', 'admin_level', 'active', 'reset_folder_count'): ('users',),
    ('mr_accept', 'mr_accept_conflicts', 'mr_accept_recurring', 'hook_archive', 'unhook_archive'): ('users',),
    ('ooo_active', 'ooo_clear', 'ooo_subject', 'ooo_message', 'ooo_from', 'ooo_until'): ('users',),
    ('add_feature', 'remove_feature', 'add_delegate', 'remove_delegate', 'add_permission', 'remove_permission'): ('users',),
    ('add_user', 'remove_user'): ('groups',),
    ('quota_override', 'quota_hard', 'quota_soft', 'quota_warn'): ('global', 'companies', 'users'),
    ('create_store', 'unhook_store', 'hook_store'): ('global', 'companies', 'users'),
    ('add_userquota_recipient', 'remove_userquota_recipient'): ('companies',),
    ('add_companyquota_recipient', 'remove_companyquota_recipient'): ('global', 'companies',),
    ('admin', 'add_view', 'remove_view', 'add_admin', 'remove_admin'): ('companies',),
}

def _encode(s):
    return s.encode(sys.stdout.encoding or 'utf8')

def orig_option(o):
    OBJ_OPT = {'users': '--user', 'groups': '--group', 'companies': '--company'}
    return OBJ_OPT.get(o) or '--'+o.replace('_', '-')

def yesno(x):
    return 'yes' if x else 'no'

def list_users(intro, users):
    users = list(users)
    print(intro + ' (%d):' % len(users))
    fmt = '{:>16}{:>20}{:>20}{:>40}'
    print(fmt.format('User', 'Full Name', 'Homeserver', 'Store'))
    print(96*'-')
    for user in sorted(users, key=lambda u: u.name):
        print(fmt.format(_encode(user.name), _encode(user.fullname), _encode(user.home_server), user.store.guid if user.store else ''))
    print

def list_groups(intro, groups):
    groups = list(groups)
    print(intro + ' (%d):' % len(groups))
    fmt = '\t{:>16}'
    print(fmt.format('Groupname'))
    print('\t'+16*'-')
    for group in sorted(groups, key=lambda g: g.name):
        print(fmt.format(_encode(group.name)))
    print

def list_companies(intro, companies):
    companies = list(companies)
    print(intro + ' (%d):' % len(companies))
    fmt = '\t{:>32}{:>32}'
    print(fmt.format('Companyname', 'System Administrator'))
    print('\t'+64*'-')
    for company in sorted(companies, key=lambda c: c.name):
        print(fmt.format(_encode(company.name), _encode(company.admin.name) if company.admin else ''))

def list_orphans(server):
    print 'Stores without users:'
    fmt = '\t{:>32}{:>20}{:>16}{:>16}{:>16}'
    print(fmt.format('Store guid', 'Username', 'Last login', 'Store size', 'Store type'))
    print('\t'+100*'-')
    for store in server.stores():
        if store.orphan:
            username = store.user.name if store.user else ''
            storesize = '%.2f MB' % (float(store.size)/2**20)
            storetype = 'public' if store.public else 'private' # XXX archive
            print(fmt.format(store.guid, _encode(username), '<unknown>', storesize, storetype))
    print
    users = [user for user in server.users() if not user.store]
    list_users('Users without stores', users)

def user_counts(server): # XXX allowed/available
    stats = server.table(PR_EC_STATSTABLE_SYSTEM).dict_(PR_DISPLAY_NAME, PR_EC_STATS_SYSTEM_VALUE)
    print 'User counts:'
    fmt = '\t{:>12}{:>10}'
    print(fmt.format('', 'Used'))
    print('\t'+42*'-')
    print(fmt.format('Active', stats['usercnt_active']))
    print(fmt.format('Non-active', stats['usercnt_nonactive']))
    print(fmt.format('NA Users', stats['usercnt_na_user']))
    print(fmt.format('NA Rooms', stats['usercnt_room']))
    print(fmt.format('NA Equipment', stats['usercnt_equipment']))
    print(fmt.format('Total', int(stats['usercnt_active'])+int(stats['usercnt_nonactive'])))

def user_details(user):
    print('Username:\t' + _encode(user.name))
    print('Fullname:\t' + _encode(user.fullname))
    print('Emailaddress:\t' + user.email)
    print('Active:\t\t' + yesno(user.active))
    print('Administrator:\t' + yesno(user.admin) + (' (system)' if user.admin_level == 2 else ''))
    print('Address Book:\t' + ('Hidden' if user.hidden else 'Visible'))
    print('Features:\t' + '; '.join(user.features))

    if user.store:
        print('Store:\t\t' + user.store.guid)
        print('Store size:\t%.2f MB' % (user.store.size / 2**20))

        if user.archive_store:
            print('Archive store:\t' + user.archive_store.guid)
        if user.archive_folder:
            print('Archive folder:\t' + user.archive_folder.path)

        print('Send-as:\t' + ', '.join(_encode(sendas.name) for sendas in user.send_as()))
        print('Delegation:\t' + ', '.join(_encode(dlg.user.name) for dlg in user.delegations()))
        print('Auto-accept meeting req:\t' + yesno(user.autoaccept.enabled))
        if user.autoaccept.enabled:
            print('Decline dbl meetingreq:\t' + yesno(not user.autoaccept.conflicts))
            print('Decline recur meet.req:\t' + yesno(not user.autoaccept.recurring))

        ooo = 'disabled'
        if user.outofoffice.enabled:
            ooo = user.outofoffice.period_desc + (' (currently %s)' % ('active' if user.outofoffice.active else 'inactive'))
        print('Out-Of-Office:\t' + ooo)

    print('Current user store quota settings:')
    print(' Quota overrides:\t' + yesno(not user.quota.use_default))
    print(' Warning level:\t\t' + str(user.quota.warning_limit or 'unlimited'))
    print(' Soft level:\t\t' + str(user.quota.soft_limit or 'unlimited'))
    print(' Hard level:\t\t' + str(user.quota.hard_limit or 'unlimited'))

    list_groups('Groups', user.groups())

    if user.store:
        print('Permissions:')
        for perm in user.permissions():
            if perm.rights: # XXX pyko remove ACE if empty
                print(_encode(' (store): ' + perm.member.name + ':' + ','.join(perm.rights)))
        for delegate in user.delegations(): # XXX merge all rights into permissions()?
            if delegate.see_private:
                print(_encode(' (store): ' + delegate.user.name + ':see_private'))
        for folder in user.folders():
            for perm in folder.permissions():
                if perm.rights:
                    print(_encode(' ' + folder.path + ': ' + perm.member.name + ':' + ','.join(perm.rights)))

def group_details(group):
    print('Groupname:\t' + _encode(group.name))
    print('Fullname:\t' + _encode(group.fullname))
    print('Emailaddress:\t' + group.email)
    print('Address Book:\t' + ('Hidden' if group.hidden else 'Visible'))
    print('Send-as:\t' + ', '.join(_encode(sendas.name) for sendas in group.send_as()))

    list_users('Users', group.users())

def company_details(company, server):
    print('Companyname:\t' + _encode(company.name))
    if company.admin:
        print('Admin:\t\t' + _encode(company.admin.name))
    print('Address Book:\t' + ('Hidden' if company.hidden else 'Visible'))
    if company.public_store:
        print('Public store:\t\t' + company.public_store.guid)
        print('Public store size:\t%.2f MB' % (company.public_store.size / 2**20))
    if server.multitenant:
        print('Remote-admin list:\t' + ', '.join(_encode(u.name) for u in company.admins()))
        print('Remote-view list:\t' + ', '.join(_encode(c.name) for c in company.views()))
        user = company.users().next()
        print('Userquota-recipient list:\t' + ', '.join(_encode(u.name) for u in user.quota.recipients() if u.name != 'SYSTEM'))
        print('Companyquota-recipient list:\t' + ', '.join(_encode(u.name) for u in company.quota.recipients() if u.name != 'SYSTEM'))

def shared_options(obj, options, server):
    if options.name:
        obj.name = options.name
    if options.fullname:
        obj.fullname = options.fullname
    if options.email:
        obj.email = options.email

    for sendas in options.add_sendas:
        obj.add_send_as(server.user(sendas))
    for sendas in options.remove_sendas:
        obj.remove_send_as(server.user(sendas))

def quota_options(user, options):
    if options.quota_override is not None:
        user.quota.use_default = not options.quota_override
    if options.quota_warn is not None:
        user.quota.warning_limit = options.quota_warn
    if options.quota_soft is not None:
        user.quota.soft_limit = options.quota_soft
    if options.quota_hard is not None:
        user.quota.hard_limit = options.quota_hard

def user_options(name, options, server):
    if options.create:
        server.create_user(name)
    user = server.user(name)

    shared_options(user, options, server)

    if options.create_store:
        user.create_store()
    if options.unhook_store:
        user.unhook()
    if options.hook_store:
        user.hook(server.store(options.hook_store))
    if options.unhook_archive:
        user.unhook_archive()
    if options.hook_archive:
        user.hook_archive(server.store(options.hook_archive))

    if options.reset_folder_count:
        for folder in [user.root] + list(user.folders()):
            folder.recount()

    if options.password:
        user.password = options.password
    if options.password_prompt:
        user.password = getpass.getpass("Password for '%s': " % user.name)
    if options.admin_level is not None:
        user.admin_level = options.admin_level
    if options.active is not None:
        user.active = options.active

    if options.mr_accept is not None:
        user.autoaccept.enabled = options.mr_accept
    if options.mr_accept_conflicts is not None:
        user.autoaccept.conflicts = options.mr_accept_conflicts
    if options.mr_accept_recurring is not None:
        user.autoaccept.recurring = options.mr_accept_recurring

    for feature in options.add_feature:
        user.add_feature(feature)
    for feature in options.remove_feature:
        user.remove_feature(feature)

    quota_options(user, options)

    for delegate in options.add_delegate:
        user.delegation(server.user(delegate), create=True)
    for delegate in options.remove_delegate:
        user.delete(user.delegation(server.user(delegate)))

    def member_rights(perm): # XXX groups
        username, rights = perm.split(':')
        rights = rights.split(',')
        see_private = 'see_private' in rights
        rights = [r for r in rights if r != 'see_private']
        return server.user(username), rights, see_private
    if (options.add_permission or options.remove_permission) and not user.store:
        raise Exception("user '%s' has no store" % user.name)
    if options.folders:
        objs = [user.folder(f) for f in options.folders]
    else:
        objs = [user.store]
    for perm in options.add_permission:
        user2, rights, see_private = member_rights(perm)
        for obj in objs:
            perm = obj.permission(user2, create=True)
            perm.rights += rights
        if see_private:
            user.delegation(user2).see_private = True
    for perm in options.remove_permission:
        user2, rights, see_private = member_rights(perm)
        for obj in objs:
            perm = obj.permission(user2)
            perm.rights = [r for r in perm.rights if r not in rights]
        if see_private:
            user.delegation(user2).see_private = False

    if options.ooo_active is not None:
        user.outofoffice.enabled = options.ooo_active
    if options.ooo_clear:
        user.outofoffice.subject = None
        user.outofoffice.message = None
        user.outofoffice.start = None
        user.outofoffice.end = None
    if options.ooo_subject is not None:
        user.outofoffice.subject = options.ooo_subject
    if options.ooo_message is not None:
        user.outofoffice.message = file(options.ooo_message).read()
    if options.ooo_from is not None:
        user.outofoffice.start = options.ooo_from
    if options.ooo_until is not None:
        user.outofoffice.end = options.ooo_until

    if options.details:
        user_details(user)
    if options.delete:
        server.delete(user)

def group_options(name, options, server):
    if options.create:
        server.create_group(name)
    group = server.group(name)

    shared_options(group, options, server)

    if options.list_users:
        list_users('User list for %s' % group.name, group.users())
    for user in options.add_user:
        group.add_user(server.user(user))
    for user in options.remove_user:
        group.remove_user(server.user(user))

    if options.details:
        group_details(group)
    if options.delete:
        server.delete(group)

def company_update_options(company, options, server):
    if options.create_store:
        company.create_public_store()
    if options.unhook_store:
        company.unhook_public_store()
    if options.hook_store:
        company.hook_public_store(server.store(options.hook_store))

    shared_options(company, options, server)

    if options.admin:
        company.admin = server.user(options.admin)

    for user in options.add_admin:
        company.add_admin(server.user(user))
    for user in options.remove_admin:
        company.remove_admin(server.user(user))
    for view in options.add_view:
        company.add_view(server.company(view))
    for view in options.remove_view:
        company.remove_view(server.company(view))

    for user in options.add_userquota_recipient:
        company.quota.add_recipient(server.user(user))
    for user in options.remove_userquota_recipient:
        company.quota.remove_recipient(server.user(user))
    for user in options.add_companyquota_recipient:
        company.quota.add_recipient(server.user(user), company=True)
    for user in options.remove_companyquota_recipient:
        company.quota.remove_recipient(server.user(user), company=True)

    for user in company.users(): # there are only server-wide settings
        quota_options(user, options)

def company_overview_options(company, options, server):
    if options.list_users:
        list_users('User list for %s' % company.name, company.users(system=True))
    if options.list_groups:
        list_groups('Group list for %s' % company.name, company.groups())

    if options.details:
        company_details(company, server)

def company_options(name, options, server):
    if options.create:
        server.create_company(name)
    company = server.company(name)

    company_update_options(company, options, server)
    company_overview_options(company, options, server)

    if options.delete:
        server.delete(company)

def global_options(options, server):
    if options.lang:
        locale.setlocale(locale.LC_MESSAGES, options.lang)

    if options.sync:
        server.sync_users()
    if options.clear_cache:
        server.clear_cache()
    if options.purge_softdelete is not None:
        server.purge_softdeletes(options.purge_softdelete)
    if options.purge_deferred:
        remaining = server.purge_deferred()
        print('Remaining deferred records: %d' % remaining)

    if options.remove_store:
        server.delete(server.store(options.remove_store))

    if options.list_orphans:
        list_orphans(server)
    if options.list_companies:
        list_companies('Company list', server.companies())
    if options.user_count:
        user_counts(server)

    if not (options.companies or options.groups or options.users):
        for company in server.companies(parse=False):
            company_overview_options(company, options, server)
            print
        company_update_options(server.company('Default'), options, server)

def check_options(options, server):
    objtypes = [name for name in ('companies', 'groups', 'users') if getattr(options, name)]
    if len(objtypes) > 1:
        raise Exception('cannot combine options: %s' % ', '.join(orig_option(o) for o in objtypes))

    actions = [name for name in sum(ACTION_MATRIX, ()) if getattr(options, name) is not None]
    if len(actions) > 1:
        raise Exception('cannot combine options: %s' % ', '.join(orig_option(a) for a in actions))

    options.details = False
    updates = [name for name in sum(UPDATE_MATRIX, ()) if getattr(options, name) not in (None, [])]
    if not (actions or updates):
        options.details = True

    for opts, legaltypes in ACTION_MATRIX.items() + UPDATE_MATRIX.items():
        for opt in opts:
            if getattr(options, opt) not in (None, []):
                illegal = set(objtypes)-set(legaltypes)
                if illegal:
                    raise Exception('cannot combine options: %s' % (orig_option(opt) + ', '+orig_option(illegal.pop())))
                if not objtypes and not 'global' in legaltypes:
                    raise Exception('%s option requires %s' % (orig_option(opt), ' or '.join(orig_option(o) for o in legaltypes)))

    if server.multitenant and not (options.companies or options.groups or options.users):
        for opt in ('create_store', 'unhook_store', 'hook_store'):
            if getattr(options, opt) is not None:
                raise Exception('%s option requires --company for multitenant setup' % orig_option(opt))

def main(options):
    server = kopano.Server(options)
    check_options(options, server)

    global_options(options, server)
    for c in options.companies:
        company_options(c, options, server)
    for g in options.groups:
        group_options(g, options, server)
    for u in options.users:
        user_options(u, options, server)

if __name__ == '__main__':
    try:
        parser, options, args = parser_opt_args()
        if args:
            raise Exception("extra argument '%s' specified" % args[0])
        main(options)
    except Exception as e:
        if 'options' in locals() and options.debug:
            print(traceback.format_exc(e))
        else:
            print(_encode(str(e)))
